### ISSUE TITLE
Enhancement: improving SE rendering

### ISSUE THREAD

**Issue by chrismaliszewski** [NPBruce/valkyrie#1128]
While testing v2.3.02b on S8+ I've noticed major delays and other issues while interacting with the SE. I think it'll be a general problem with the current version of SE. 

1. By "delays" I mean: on PC the list is generated almost immediately, <1s, while on a smartphone it takes ~2.5-4s. Of course, the performance difference of both types of devices is unquestionable but a generation of a simple list shouldn't take that long. 
2. What is worse, the list is generated from scratch every time when you open Select Scenario screen. 
3. (I wrote it on another ticket but can't find it) After you go back from a scenario screen, the list is generated from scratch instead of going back to the point you were with the scenario.
4. While the list is generated you can't scroll as it's populated with new elements and refreshed preventing the action. It's a little iritating when you try to scroll but you can't and you're returned to the top of the list.

I find 2-4 especially annoying. 

I don't know Unity so I can't say if the below propositions are possible.
1. I have an idea to speed up the process: caching or layering the screen. Not sure if it's possible but if 'yes', the SE screen could be created in the background or as a separate layer hidden at the bottom and moved to the front when needed. **One more thing:** download and save scenario's images on the device so data is not used each time you run the app again. 
2. Read above with a small difference. Instead of generating the list in the background, the screen is hidden in the background layer. It's removed from memory when a user goes to Main Menu or starts a scenario. Using this method, when a user opens a scenario's splash-screen and goes back, the list is already created in the background...
3. The 2. resolution also solved this problem. As the list is already created, a user goes back directly from where he/she left. If caching is not possible, another idea is possible: block the list until it's fully generated, i.e.
![image](https://user-images.githubusercontent.com/7532602/54877193-8ee58200-4e1b-11e9-88b4-a5ca03918138.png)
4. The above is both solved by caching/layering and the resolution from 3, blocking the list until it's fully generated.

**Comment by BenRQ** [comment]
I agree that some loading times are still very frustrating, especially on Android.

I would say that the scenario explorer is not the worse though.

I would rate it in that order, from worse to 'bad but not too bad' : 
1) Starting a big scenario from the scenario details screen (file is unzipping)
2) initial loading time, after selecting the game (MoM vs Descent), this can be slow when you have savegames of big scenarios
3) very first loading of scenario explorer, when image are not cached
4) second loading of scenario explorer, when image are cached

In all cases, adding a (moving) loading screen would certainly makes thing better.

I would not cache the whole scenario explorer as it's going to take a lot of memory, and could be problematic for devices.

And FYI I already did a lot of optimization in the scenario explorer.
Images are cached.

**Comment by chrismaliszewski** [comment]
Hi @BenRQ. I want you to know that I **really** appreciate and admire the change made in v2.3 with the SE! It's a huge improvement and a great job done! 

Being honest, I don't have any problems with loading times in general at all but in SE. Points 1. and 2., at least for me, aren't the issue. 

But, could you elaborate about 2.? What does the initial menu load have anything to do with the save games? I can't think what you mean. If you have time, of course. 

Regarding caching images; caching works only during a single running session. If you turn off the app and turn it on again, the images seem to be lazily loaded as they aren't displayed immediately comparing to the second opening of SE after being cached. I'm talking about the kind of caching images. 

Before I write my proposition (which may or may not be good), could you please tell me something:
1. What format has the data downloaded from the server; is it a single file which contains information about all scenarios or is it a bunch of files separate for each scenario?
2. The data which is downloaded is the manifest I read about many times?
3. Where does sorting take place; on the server or on the client's side?
4. Are the files stored/cached somewhere on the client's side or are they downloaded fully each time a user opens SE?
That should clarify many things :).

**Comment by BenRQ** [comment]
> But, could you elaborate about 2.? What does the initial menu load have anything to do with the save games? I can't think what you mean. If you have time, of course.

The game menu checks if there are savegames, and if they are valid.
If not, the load button is made grey and you can't click it.
It's just the way it is, one of the many things that needs to be improved.

Caching image on local disk is probably not a big priority, images are loaded after the list is fully loaded and available. So you can navigate in it, already.

1) 2) It is a single file, compiled automatically by a script every hour.
https://drive.google.com/file/d/13JEtzRQ1LcCAAhKluxii0tgKDW71XODV/view

3) Sorting is done on client side, but time to sort 40 data is really negligible

4) The one single file is downloaded once by the app during runtime.
If you restart the app you will download it again.
We are talking about a few kilobyte here, nothing big :)

What is taking time is the creation of graphical elements in Unity.
There are a lot of elements for each scenario, and I already cleaned up a lot of the code.

**Comment by chrismaliszewski** [comment]
I created a ticket regarding saves: https://github.com/NPBruce/valkyrie/issues/1130
And another ticket about caching: https://github.com/NPBruce/valkyrie/issues/1131

After reading
>What is taking time is the creation of graphical elements in Unity.

I remember having a similar discussion in the past, perhaps even with you :). The answers "obstructed" the idea since it's Unity "issue". 

Nevertheless, I checked QuestSelectionScreen.cs. Now I see what causes such long loading time and problems with scrolling. I'm only not sure how scrollArea, frame and ui are connected? Is it something like:
![image](https://user-images.githubusercontent.com/7532602/55065343-f0a42700-507b-11e9-907d-d4ea807adc63.png)
Wouldn't it be better to use https://docs.unity3d.com/Manual/script-ScrollRect.html and populate it the list elements? (example of usage: https://www.youtube.com/watch?v=ZI6DwJtjlBA and other movies) It wouldn't be you/the app who tells Unity how and where to display everything but just what to display. It might be faster. 
I was also thinking about generating images as list items and display them but there are too many things which may change (user's laguage, expansions, statistics etc.) so there isn't much sense in that approach.

**Comment by BenRQ** [comment]
> The answers "obstructed" the idea since it's Unity "issue".

I'm not saying it is a Unity issue.
We are creating a big amount of UI element so it is taking some time.
I obviously tried to limit the number of different UI element created per scenario, and I've reached a point where the time spend optimizing is not making a big change.

> I'm only not sure how scrollArea, frame and ui are connected?

```
<scrollarea>
<Scenario frame>
</variousUIElement></variousUIElement></variousUIElement></variousUIElement>
<...>
</variousUIElement></variousUIElement></variousUIElement></variousUIElement>
</Scenario frame>
<...>
<Scenario frame>
</variousUIElement></variousUIElement></variousUIElement></variousUIElement>
<...>
</variousUIElement></variousUIElement></variousUIElement></variousUIElement>
</Scenario frame>
</scrollarea>
```

As I said before, keeping all those elements in memory would take a lot of memory.

>   Wouldn't it be better to use https://docs.unity3d.com/Manual/script-ScrollRect.html and populate it the list elements?

We are already using ScrollRect for this, check UIElementScrollVertical class.
`UnityEngine.UI.ScrollRect scrollRect = scrollBG.AddComponent<UnityEngine.UI.ScrollRect>(); `


For me the reasonable solution for SE, is having a loading indicator.

**Comment by chrismaliszewski** [comment]
Just saying, your pseudo-XML reminded me of creating UI based on XML which I was doing in one of my previous jobs. But after changing the Internet, there only paid classes for that in Unity, so never mind :).

Sorry about ScrollRect. I didn't check the origin of UIElementScrollVertical, only QuestSelectionScreen.cs file. 

Ok, so... I have only one idea left which I find a good practice with lists in the web development. At the moment, you create each Frame's UIElement separately, one by one. Instead, you could initialize just one UIElement which would be your template, clone it, then make required changes in texts, images etc. and render it where it's needed. 
Normally, in web development, this approach it's much faster and cleaner in terms of coding, but the question with Unity is "how much faster would it be?"
I think the change wouldn't be hard to do as most of the pieces of code are already there. 

You're the master of Unity, not me :). What do you think?

**Comment by BenRQ** [comment]
I haven't tried much with object pooling.
It could be interesting to try this at UIElement level maybe, so it works for both SE and the whole valkyrie experience.
